lib/puppet/test/test_helper.rb

```diff
@@ -120,8 +120,11 @@ def self.before_each_test()
       indirections = Puppet::Indirector::Indirection.send(:class_variable_get, :@@indirections)
       indirections.each do |indirector|
         $saved_indirection_state[indirector.name] = {
-            :@terminus_class => indirector.instance_variable_get(:@terminus_class),
-            :@cache_class    => indirector.instance_variable_get(:@cache_class)
+          :@terminus_class => indirector.instance_variable_get(:@terminus_class),
+          :@cache_class    => indirector.instance_variable_get(:@cache_class),
+          # dup the termini hash so termini created and registered during
+          # the test aren't stored in our saved_indirection_state
+          :@termini        => indirector.instance_variable_get(:@termini).dup
```

Comment:
fwiw, it looks like indirections have a reader for this https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/indirection.rb#L14. It looks like in master terminus_class & cache_class became threadsafe but not termini. A merge up will be a little complicated because how we set these values won't work quite the same now: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/test/test_helper.rb#L179 .
It looks like in 5.5.x and master we have reader and writers for these three values which will do the right thing before and after the threadsafety work, might make the merge up easier (and will invalidate some of pittman's comment above) to use them instead of instance_variable_{get,set}.

Reply:
I couldn't use the `terminus_class` reader because it performs additional validation if the current value is nil, raising if a setting doesn't specify the default terminus. I'll submit a separate PR for the merge up.

Also since each terminus is [lazily loaded](https://github.com/puppetlabs/puppet/blob/70f3050f437db45b374c084fb3361d57d7e81706/lib/puppet/indirector/indirection.rb#L140) should the call to `make_terminus` be protected with a lock? Or is it safe to let them race and have the last one win? I assume the instance loader is safe due to the concurrent locks @pcarlisle?

Reply:
I would probably lean to having a lock there especially since something might have a side effect when loaded (though hopefully not)

Reply:
We only call that unsafe code when files are read initially, or when routes are configured, both should happen when Puppet server's JRuby is still single threaded. But I suppose a user could load a termini within their own function or something that's called during evaluation or report processing? It's probably safest to put a lock on it. I can open a ticket for that.

I didn't realize setting a termini_class to nil was a thing tests did, that's a shame.

Reply:
Yeah, sorry, I think I'm way wrong about that above comment. Here's the ticket though: https://tickets.puppetlabs.com/browse/PUP-10594